### PR TITLE
satellite/accounting: add redis support to live accounting

### PR DESCRIPTION
**Pull request by cam-a**
What: add support for redis backend for live accounting cache
https://storjlabs.atlassian.net/browse/V3-1836

Why: So multiple servers can access live accounting data

Please describe the tests:
 - Test 1: test that redis backend works
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/accounting/live/redis.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/skyrings/skyring-common/tools/uuid"
+	"go.uber.org/zap"
+
+	"storj.io/storj/storage"
+	"storj.io/storj/storage/redis"
```
Comment (egonelbre) [inline satellite/accounting/live/redis.go]: Avoid direct dependencies to a database implementation in satellite services etc. The reason is to make it easy to add new database implementation, without requiring it to be a dependency to a specific binary. This probably is less important to satellite than to storage node and uplink, but still a good practice.

File: satellite/accounting/live/live.go [hunk 2]
```diff
@@ -25,7 +24,12 @@ type Config struct {
 type Service interface {
 	GetProjectStorageUsage(ctx context.Context, projectID uuid.UUID) (int64, int64, error)
 	AddProjectStorageUsage(ctx context.Context, projectID uuid.UUID, inlineSpaceUsed, remoteSpaceUsed int64) error
-	ResetTotals()
+	ResetTotals() error
```
Comment (egonelbre) [inline satellite/accounting/live/live.go]: `context` is missing

File: satellite/accounting/live/live_test.go [hunk 3]
```diff
@@ -14,15 +14,74 @@ import (
...
+	for _, projID := range projectIDs {
+		inlineUsed, remoteUsed, err := service.GetProjectStorageUsage(ctx, projID)
+		require.NoError(t, err)
+		assert.Equalf(t, sum, inlineUsed, "projectID %v", projID)
+		assert.Equalf(t, sum, remoteUsed, "projectID %v", projID)
+	}
+}
+
+func TestRedisLiveAccounting(t *testing.T) {
+	ctx := testcontext.New(t)
```
Comment (egonelbre) [inline satellite/accounting/live/live_test.go]: testcontext cleanup is missing.

File: satellite/accounting/live/live_test.go [hunk 4]
```diff
@@ -14,15 +14,74 @@ import (
...
+	"storj.io/storj/storage/redis/redisserver"
 )
 
 func TestPlainMemoryLiveAccounting(t *testing.T) {
-	const (
-		valuesListSize  = 1000
-		valueMultiplier = 4096
-		numProjects     = 200
-	)
+	ctx := testcontext.New(t)
```
Comment (egonelbre) [inline satellite/accounting/live/live_test.go]: testcontext cleanup is missing.

File: satellite/accounting/live/redis.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		return 0, 0, err
+	}
+	return curVal.InlineSpace, curVal.RemoteSpace, nil
+}
+
+// AddProjectStorageUsage lets the live accounting know that the given
+// project has just added InlineSpaceUsed bytes of inline space usage
+// and RemoteSpaceUsed bytes of remote space usage.
+func (cache *redisLiveAccounting) AddProjectStorageUsage(ctx context.Context, projectID uuid.UUID, InlineSpaceUsed, RemoteSpaceUsed int64) error {
+	curInlineTotal, curRemoteTotal, err := cache.GetProjectStorageUsage(ctx, projectID)
```
Comment (egonelbre) [inline satellite/accounting/live/redis.go]: Add tests for the concurrent calls.

For example, let's say you do `100` concurrent calls to `AddProjectStorageUsage` then the result should be the total.

File: satellite/accounting/live/redis.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	if err != nil {
+		return 0, 0, err
+	}
+	return curVal.InlineSpace, curVal.RemoteSpace, nil
+}
+
+// AddProjectStorageUsage lets the live accounting know that the given
+// project has just added InlineSpaceUsed bytes of inline space usage
+// and RemoteSpaceUsed bytes of remote space usage.
+func (cache *redisLiveAccounting) AddProjectStorageUsage(ctx context.Context, projectID uuid.UUID, InlineSpaceUsed, RemoteSpaceUsed int64) error {
```
Comment (egonelbre) [inline satellite/accounting/live/redis.go]: Start arguments with lower case.

File: satellite/accounting/live/live.go (hunk 2 above)
Comment (egonelbre) [reply satellite/accounting/live/live.go]: Yeah.

File: satellite/accounting/live/redis.go (hunk 1 above)
Comment (cam-a) [reply satellite/accounting/live/redis.go]: I think I see what you mean. Pass a storage.KeyValueStore into the service? But what should I do with the `plainMemoryLiveAccounting` that was already there, since it doesn't match the KeyValueStore interface?

Comment (egonelbre) [reply satellite/accounting/live/redis.go]: You would pass a specific `Service` or `DB` rather than `storage.KeyValueStore`... The key value is one possible implementation, not the definitive.

File: satellite/accounting/live/redis.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		return err
+	}
+	return cache.client.Put(ctx, []byte(projectID.String()), marshalled)
+}
+
+// ResetTotals reset all space-used totals for all projects back to zero. This
+// would normally be done in concert with calculating new tally counts in the
+// accountingDB.
+func (cache *redisLiveAccounting) ResetTotals(ctx context.Context) error {
+	cache.log.Info("Resetting real-time accounting data")
```
Comment (ifraixedes) [inline satellite/accounting/live/redis.go]: Change this log to `Debug` level

File: satellite/accounting/live/redis.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	return cache.client.FlushDB()
+}
+
+// Close the DB connection
```
Comment (ifraixedes) [inline satellite/accounting/live/redis.go]:
```suggestion
// Close the DB connection.
```

File: satellite/accounting/live/plainmemory.go [hunk 9]
```diff
@@ -76,28 +40,32 @@ func (pmac *plainMemoryLiveAccounting) GetProjectStorageUsage(ctx context.Contex
...
 	pmac.spaceDeltas[projectID] = curVal
 	return nil
 }
 
 // ResetTotals reset all space-used totals for all projects back to zero. This
 // would normally be done in concert with calculating new tally counts in the
 // accountingDB.
-func (pmac *plainMemoryLiveAccounting) ResetTotals() {
+func (pmac *plainMemoryLiveAccounting) ResetTotals(ctx context.Context) error {
 	pmac.log.Info("Resetting real-time accounting data")
```
Comment (ifraixedes) [inline satellite/accounting/live/plainmemory.go]: I would change this to a `Debug` in this branch although this line has not been modified in this branch.

File: satellite/accounting/live/live_test.go [hunk 10]
```diff
@@ -14,35 +14,188 @@ import (
...
+
+	config := Config{
+		StorageBackend: "plainmemory",
+	}
+	cache, err := NewCache(zaptest.NewLogger(t).Named("live-accounting"), config)
+	require.NoError(t, err)
+
+	// ensure we are using the expected underlying type
+	_, ok := cache.(*plainMemoryLiveAccounting)
+	assert.True(t, ok)
```
Comment (ifraixedes) [inline satellite/accounting/live/live_test.go]: Why not using [`IsType`](https://godoc.org/github.com/stretchr/testify/assert#IsType)?

File: satellite/accounting/live/live_test.go [hunk 11]
```diff
@@ -14,35 +14,188 @@ import (
...
+
+	config := Config{
+		StorageBackend: "redis://" + address + "?db=0",
+	}
+	cache, err := NewCache(zaptest.NewLogger(t).Named("live-accounting"), config)
+	require.NoError(t, err)
+
+	// ensure we are using the expected underlying type
+	_, ok := cache.(*redisLiveAccounting)
+	assert.True(t, ok)
```
Comment (ifraixedes) [inline satellite/accounting/live/live_test.go]: Why not using [`IsType`](https://godoc.org/github.com/stretchr/testify/assert#IsType)?

File: satellite/accounting/live/live_test.go [hunk 12]
```diff
@@ -14,35 +14,188 @@ import (
...
 	}
-	service, err := New(zaptest.NewLogger(t).Named("live-accounting"), config)
+	cache, err := NewCache(zaptest.NewLogger(t).Named("live-accounting"), config)
 	require.NoError(t, err)
 
 	// ensure we are using the expected underlying type
-	_, ok := service.(*plainMemoryLiveAccounting)
-	require.True(t, ok)
+	_, ok := cache.(*plainMemoryLiveAccounting)
+	assert.True(t, ok)
```
Comment (ifraixedes) [inline satellite/accounting/live/live_test.go]: Why not using [`IsType`](https://godoc.org/github.com/stretchr/testify/assert#IsType)?

File: satellite/accounting/live/live_test.go [hunk 13]
```diff
@@ -14,35 +14,188 @@ import (
...
+	}
+	require.NoError(t, group.Wait())
+
+	spaceUsed, err := cache.GetProjectStorageUsage(ctx, projectID)
+	require.NoError(t, err)
+
+	require.EqualValues(t, expectedSum, spaceUsed)
+}
+
+func TestResetTotals(t *testing.T) {
```
Comment (ifraixedes) [inline satellite/accounting/live/live_test.go]: I'm wondering if this name is appropriated for this test. It's called `ResetTotals` but I don't see the call to it.

Comment (cam-a) [reply satellite/accounting/live/live_test.go]: Yes, I'm not sure what this test is for

Comment (ifraixedes) [reply satellite/accounting/live/live_test.go]: It looks that this function is already tested in the `Redis` and `Plainmemory` tests, so we should remove this one, unless that we want a test which passes a negative number to `AddProjectStorageUsage` (https://github.com/storj/storj/pull/3213/files/6ba21dabef6135a3645541d4d7dfa8e76ddf22ed#diff-72f4bb03961ddf4ea47393d293e27c2fR180), in that case, we should rename it.

Comment (cam-a) [reply satellite/accounting/live/live_test.go]: Hmm, looking at the test now, I see that it is actually possible to subtract from the space used by passing a negative number into `AddProjectStorageUsage`, which I don't think we want to do. I'll add some checks and use this test to ensure that we can't pass negative numbers into the function

File: satellite/accounting/projectusage.go [hunk 14]
```diff
@@ -118,25 +117,25 @@ func (usage *ProjectUsage) ExceedsStorageUsage(ctx context.Context, projectID uu
 	}
 
 	maxUsage := limit.Int64() * int64(ExpansionFactor)
-	if inlineTotal+remoteTotal >= maxUsage {
+	if totalUsed >= maxUsage {
```
[the remark this answers is not in the document]
Comment (cam-a) [reply satellite/accounting/projectusage.go]: I believe this function is used when creating a new segment and if `totalUsed == maxUsage`,  it will return an error and the segment will not be committed because the limit has been reached

File: satellite/accounting/live/cache.go [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+func NewCache(log *zap.Logger, config Config) (accounting.LiveAccounting, error) {
+	parts := strings.SplitN(config.StorageBackend, ":", 2)
+	var backendType string
+	if len(parts) == 0 || parts[0] == "" {
+		backendType = "plainmemory"
+	} else {
+		backendType = parts[0]
+	}
+	switch backendType {
+	case "plainmemory":
```
Comment (egonelbre) [inline satellite/accounting/live/cache.go]: nit: this could just be memory, like sqlite uses

File: satellite/accounting/live/live_test.go [hunk 16]
```diff
@@ -5,6 +5,7 @@ package live
```
Comment (egonelbre) [inline satellite/accounting/live/live_test.go]: nit: tests usually should be using `_test` suffix to ensure it's not relying on non-exposed behavior

File: satellite/accounting/live/live_test.go [hunk 17]
```diff
@@ -14,35 +15,190 @@ import (
 	"go.uber.org/zap/zaptest"
 	"golang.org/x/sync/errgroup"
 
+	"storj.io/storj/internal/testcontext"
 	"storj.io/storj/internal/testrand"
+	"storj.io/storj/satellite/accounting"
+	"storj.io/storj/storage/redis/redisserver"
 )
 
 func TestPlainMemoryLiveAccounting(t *testing.T) {
```
Comment (egonelbre) [inline satellite/accounting/live/live_test.go]: nit: you can write one testsuite that tests both implementations rather than writing tests for each

File: satellite/accounting/live/plainmemory.go [hunk 18]
```diff
@@ -56,48 +23,57 @@ type plainMemoryLiveAccounting struct {
...
-	remoteSpace int64
+	spaceDeltas  map[uuid.UUID]int64
 }
 
 func newPlainMemoryLiveAccounting(log *zap.Logger) (*plainMemoryLiveAccounting, error) {
 	pmac := &plainMemoryLiveAccounting{log: log}
-	pmac.ResetTotals()
+	pmac.spaceMapLock.Lock()
+	pmac.spaceDeltas = make(map[uuid.UUID]int64, 0)
+	pmac.spaceMapLock.Unlock()
```
Comment (egonelbre) [inline satellite/accounting/live/plainmemory.go]: there's no need for the lock here since nothing else can be holding a reference to it.

File: satellite/accounting/live/redis.go [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+	"go.uber.org/zap"
+
+	"storj.io/storj/storage"
+	"storj.io/storj/storage/redis"
+)
+
+type redisLiveAccounting struct {
+	log *zap.Logger
+
+	client *redis.Client
```
Comment (egonelbre) [inline satellite/accounting/live/redis.go]: this package is still depending on a redis, but this can be separated also in a different PR

File: satellite/api.go [hunk 20]
```diff
@@ -96,7 +95,7 @@ type API struct {
 	}
 
 	LiveAccounting struct {
-		Service live.Service
+		Cache accounting.LiveAccounting
```
Comment (egonelbre) [inline satellite/api.go]: Why not name it `Cache` or `TotalCache` or similar?
